### DISCUSSION TITLE
Creating smart contract without Hedera File Service

### DISCUSSION THREAD

**Discussion by wensheng** [hiero-ledger/hiero-improvement-proposals#246] [Ideas]
Abstract
======
User should be able to create smart contract on Hedera by uploading bytecode directly without first uploading it to Hedera File Service (HFS).

Motivation
========
At present, to create smart contract on Hedera, one need to upload smart contract bytecode to HFS, get a a FileID back.  If the bytecode is large, the user additionally has to perform FileAppendTransaction.  Then the user can create smart contract (with `setBytecodeFileId`) using the returned FileId.  This poses usability issue and hamper the adoption of Hedera smart contract.

Rationale
=======
1. The smart contract contains bytecode, otherwise it will not function, storing bytecode somewhere else is not needed.
2. It's much easier and more convenient for user to just upload bytecode directly to create smart contract.
3. Right now, after the contract is created, the file on HFS can be deleted without affecting smart contract.  This further enhance the argument that HFS is not needed for smart contract. 

User stories
=========
1. As a smart contract developer, I want to create a smart contract on Hedera by uploading bytecode directly.  I understand it may take a little longer and gas fee may be higher, compared with using HFS.  (But reasonably it should be faster and cheaper to create smart contract from bytecode directly, because it's one transaction vs 2 or 3 transactions)

**Comment by shemnon** [comment]
The file service overcomes the 6KiB signed transaction size limit.  Some smart contracts require 48KiB of init code. So we cannot get rid of the file service use completely.

**Comment by 0xJepsen** [comment]
Perhaps the way to resolve this would be to abstract to file deploying from the end developer, so they only have one API call rather than many. This way, you can still bypass the 6KiB transaction size and still deliver a better developer experience.

**Comment by gregscullard** [comment]
The file service is a necessary step, however the chain of transactions (file create, file append, contract create) can be abstracted by the SDKs. Proposed improvement raised on Java, Go and JavaScript SDKs below:

https://github.com/hashgraph/hedera-sdk-java/issues/839
https://github.com/hashgraph/hedera-sdk-go/issues/364
https://github.com/hashgraph/hedera-sdk-js/issues/850

**Comment by gregscullard** [comment]
@wensheng if you agree with the above, we can create a HIP to drive the modifications to the SDK.

**Reply by wensheng** [reply]
Sure, please go ahead.
Thank you

**Reply by 0xJepsen** [reply]
@gregscullard circling back on this, Let me know if there is anything I can do to help!

**Comment by gregscullard** [comment]
@wensheng @0xJepsen HIP created here: https://github.com/hashgraph/hedera-improvement-proposal/pull/342

**Reply by 0xJepsen** [reply]
Awesome!

**Comment by shemnon** [comment]
Bugs for this same issue but using HIP-410 facilities - 

https://github.com/hashgraph/hedera-sdk-go/issues/455
https://github.com/hashgraph/hedera-sdk-js/issues/1043
https://github.com/hashgraph/hedera-sdk-java/issues/948

**Comment by Nana-EC** [comment]
@wensheng just closing the loop here. The [ContractCreateFlow](https://docs.hedera.com/hedera/sdks-and-apis/sdks/smart-contracts/create-a-smart-contract#contractcreateflow) was implemented and has simplified the process for users.

**Comment by Neurone** [comment]
I'd like to say that in addition to the ContractCreateFlow - that still uses HFS under the hood - you can technically create a smart contract via EthereumTransactions without using HFS. As Danno already mentioned, you are limited to the maximum size of a Hedera transaction (6Kb, signatures included) but you can do it.
